src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs

```diff
@@ -0,0 +1,470 @@
+
+            var diagnosticsMock = new Mock<IDiagnosticService>();
+            // Create diagnostic for the first document that has a mapped location.
+            var mappedFilePath = documents[0].FilePath + "m1";
+            var documentOneDiagnostic = await CreateMockDiagnosticDatasWithMappedLocationAsync(documents[0], ("doc1Diagnostic", mappedFilePath)).ConfigureAwait(false);
+            // Create diagnostic for the second document that maps to the same location as the first document diagnostic.
+            var documentTwoDiagnostic = await CreateMockDiagnosticDatasWithMappedLocationAsync(documents[1], ("doc2Diagnostic", mappedFilePath)).ConfigureAwait(false);
+
+            // On the first call for the documents, return the mapped diagnostic.  On the second, return nothing.
+            SetupMockDiagnosticSequence(diagnosticsMock, documents[0].Id, documentOneDiagnostic, ImmutableArray<DiagnosticData>.Empty);
+            SetupMockDiagnosticSequence(diagnosticsMock, documents[1].Id, documentTwoDiagnostic, ImmutableArray<DiagnosticData>.Empty);
+
+            // Publish four document change diagnostic notifications ->
+            // 1.  doc1 with doc1Diagnostic = mapped file m1.
+            // 2.  doc2 with doc2Diagnostic = mapped file m1.
+            // 3.  doc1 with empty diagnostics.
+            // 4.  doc2 with empty diagnostics.
+            //
+            // We expect four publish diagnostics ->
+            // 1.  from URI m1 with doc1Diagnostic (triggered by 1 above to add doc1Diagnostic).
+            // 2.  from URI m1 with doc1Diagnostic and doc2Diagnostic (triggered by 2 above to add doc2Diagnostic).
+            // 3.  from URI m1 with just doc2Diagnostic (triggered by 3 above to clear doc1 diagnostic).
+            // 4.  from URI m1 with empty (triggered by 4 above to also clear doc2 diagnostic).
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, 4, documents[0], documents[1], documents[0], documents[1]).ConfigureAwait(false);
+
+            Assert.Equal(4, results.Count);
+            var expectedUri = new Uri(mappedFilePath);
+            Assert.Equal(expectedUri, results[0].Uri);
+            Assert.Equal("doc1Diagnostic", results[0].Diagnostics.Single().Code);
+
+            Assert.Equal(expectedUri, results[1].Uri);
+            Assert.Equal(2, results[1].Diagnostics.Length);
+            Assert.True(results[1].Diagnostics.Contains(d => d.Code == "doc1Diagnostic"));
+            Assert.True(results[1].Diagnostics.Contains(d => d.Code == "doc2Diagnostic"));
+
+            Assert.Equal(expectedUri, results[2].Uri);
+            Assert.Equal(1, results[2].Diagnostics.Length);
+            Assert.True(results[2].Diagnostics.Contains(d => d.Code == "doc2Diagnostic"));
+
+            Assert.Equal(expectedUri, results[3].Uri);
+            Assert.True(results[3].Diagnostics.IsEmpty());
+        }
```

Comment:
This probably needs some asserts here that the underling maps are empty otherwise they're potentially leaking...